### PR TITLE
[draft PR][poc] Meta queries alert

### LINKED ISSUE #15696: [Feature Request] Support mixed type plugin for backend
Trying to port https://github.com/GoshPosh/grafana-meta-queries to backend so that alerting is supported.

Mixed plugins behave differently as in they require access to all queries in the dashboard and respective datasources.

Mixed data source gets all the queries & datasource of a given dashboard:
https://github.com/grafana/grafana/commit/b30dfcf28a92d770e02c007dc67b309acb287cef
and is used like : https://github.com/GoshPosh/grafana-meta-queries/blob/master/src/datasource.js#L43

But the same thing is not achievable on backend plugins.
Backend plugins get call to 
```
func (t *DruidDatasource) Query(ctx context.Context, tsdbReq *datasource.DatasourceRequest) (*datasource.DatasourceResponse, error) {
```

which has access to current query only. Tried digging through the code 
https://github.com/grafana/grafana/blob/master/pkg/plugins/datasource/wrapper/datasource_plugin_wrapper.go#L61
sends only the current query and datasource

Digging a little further
the single query is saved within the alert table of grafana database and hence has no information to the dashboard.

Unsure how to proceed

Included information:
### What Grafana version are you using?
grafana 5 & 6
### What datasource are you using?
https://github.com/GoshPosh/grafana-meta-queries
with druid
### What OS are you running grafana on?
mac & linux
 
relates to  https://github.com/grafana/grafana/issues/6841

### PR DESCRIPTION
**Pull request by Gauravshah**
**What this PR does / why we need it**:
This plugin sends all of the queries to backend plugin instead of only sending the query on which the alert is turned on.
This enables plugins like metaqueries to be able to make additional computations or even call the underlying datasource to get datapoints from that plugin.

**Which issue(s) this PR fixes**:
Fixes #15696

**Special notes for your reviewer**:
PR is for approach review, far from mergeable state
Please ignore the dev changes, PR is meant for POC purposes only

**Approach taken**:
Move the metaqueries plugin inside grafana repo.
Make the tsdb.request send all the queries instead of only alert query.
call `tsdb.HandleRequest(` from within metaqueries to access other plugins
return back results as any other plugin

### CODE REVIEW

File: pkg/services/alerting/conditions/query.go [hunk 1]
```diff
@@ -110,6 +110,45 @@ func (c *QueryCondition) executeQuery(context *alerting.EvalContext, timeRange *
...
+				Name: dataSourcename.MustString(),
+				//OrgId: context.Rule.OrgId,
+			}
+
+			if err := bus.Dispatch(getTargetDsInfo); err != nil {
+				return nil, fmt.Errorf("Could not find datasource %v", err)
+			}
+
+			tsdbQuery := tsdb.Query{RefId: targetJSON.Get("refId").MustString(), Model: targetJSON, DataSource: getTargetDsInfo.Result}
+			req.Queries = append(req.Queries, &tsdbQuery)
```
Comment (Gauravshah) [inline pkg/services/alerting/conditions/query.go]: @torkelo  can not append all Queries of dashboard to `req.Targets` similar to JS

File: pkg/tsdb/metaqueries/metaqueries.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+				druid = tsdbQuery
+				for i := 0; i < len(druid.Queries); i++ {
+					druid.Queries = append(druid.Queries[:i], druid.Queries[i+1:]...)
+				}
+				druid.Queries[len(druid.Queries)-1] = nil
+				druid.Queries = []*tsdb.Query{}
+
+				druid.Queries = append(druid.Queries, druidQuery)
+
+				response, err = tsdb.HandleRequest(ctx, druid.Queries[0].DataSource, druid)
```
Comment (Gauravshah) [inline pkg/tsdb/metaqueries/metaqueries.go]: @bergquist I am not sure how to do this call if we move plugin outside as its own executable. Some guidance would be super helpful

**Comment by Gauravshah** [comment]
@bergquist / @torkelo did you get a chance to validate this approach ?

**Comment by Gauravshah** [comment]
ping @bergquist

**Comment by Gauravshah** [comment]
ping

**Comment by Gauravshah** [comment]
@marefr Is there something that is missing and hence this is not getting a review ?

**Comment by marefr** [comment]
@Gauravshah thank you for contributing to Grafana and sorry for the late response.

We have been struggling a bit our self for how to supporting these kind of use cases in Grafana. We've come to the conclusion that we would like to support datasource plugins processing data from other datasources, but instead expose it thru the Grafana plugin model so that several datasources/plugins could make use of this. 

This is why we think your proposed solution of putting meta queries plugin in core Grafana is not a good idea and therefore we suggest closing this PR.

We thank you for your proof of concept and it highlights a couple of tricky things to take into consideration when implementing a plugin model for this. This is very valuable for us. 

Again, thank you.

**Comment by Gauravshah** [comment]
Thanks for the response @marefr 
I definitely agree adding another plugin to core is not the right way to go.

> but instead expose it thru the Grafana plugin model so that several datasources/plugins could make use of this.

Can you give some pointers on how this is possible. Will try to work in that direction

**Comment by marefr** [comment]
@Gauravshah to early to say exactly in what direction this will take from a [Grafana plugin model](https://github.com/grafana/grafana-plugin-model) perspective. As it looks now we'll work on Grafana plugin model in 6.5 release and I have to come back later when I know more.

**Comment by Gauravshah** [comment]
Thanks @marefr will take a look at it post 6.5 release

**Comment by Gauravshah** [comment]
@marefr Were we able to get any changes for this in Grafana 6.5 release ? I was not able to find it in releases section, was unsure if these were internal

**Comment by zanybug** [comment]
@marefr could you please give us some input on how to send email notifications for Meta query alerts?

image

**Comment by Gauravshah** [comment]
ping @marefr

**Pull request closed without merging**